wpilibj/src/main/java/edu/wpi/first/wpilibj/PowerDistributionLogger.java

```diff
@@ -0,0 +1,322 @@
+ */
+public class PowerDistributionLogger {
+  private final int m_handle;
+  private final Consumer<PowerDistributionStats> m_output;
+  private final PowerDistributionStats m_stats = new PowerDistributionStats();
+
+  /**
+   * This value determines how many calls to {@link #log()} are required before the data is logged.
+   */
+  private int m_logWindup;
+
+  /** This value tracks the number of times {@link #log()} has been called. */
+  private int m_logCount;
+
+  /**
+   * Constructs a PowerDistributionLogger, this provides a performant way to log PowerDistribution
+   * data to NetworkTables or only to a DataLog.
+   *
+   * @param pd The PowerDistribution object to log
+   * @param logPath The path to log the data to
+   * @param datalogOnly If true, only log to the datalog
+   */
+  public PowerDistributionLogger(PowerDistribution pd, String logPath, boolean datalogOnly) {
+    m_handle = pd.getHandle();
+    String statsLogPath = NetworkTable.normalizeKey(logPath + "/Stats", true);
+    String metaLogPath = NetworkTable.normalizeKey(logPath + "/Meta", true);
+    PowerDistributionMeta meta = new PowerDistributionMeta(m_handle);
+    if (datalogOnly) {
+      StructLogEntry<PowerDistributionStats> entry =
+          StructLogEntry.create(
+              DataLogManager.getLog(), statsLogPath, new PowerDistributionStatsStruct());
+      m_output = entry::append;
+      StructLogEntry.create(DataLogManager.getLog(), metaLogPath, new PowerDistributionMetaStruct())
+          .append(meta);
+    } else {
+      StructPublisher<PowerDistributionStats> entry =
+          NetworkTableInstance.getDefault()
+              .getStructTopic(statsLogPath, new PowerDistributionStatsStruct())
+              .publish();
+      m_output = entry;
+      NetworkTableInstance.getDefault()
+          .getStructTopic(metaLogPath, new PowerDistributionMetaStruct())
+          .publish()
+          .set(meta);
+    }
+  }
+
+  /**
+   * Logs the PowerDistribution data to NetworkTables or DataLog.
+   *
+   * <p>If {@link #setLogWindup(int)} has been called, this method will only log every N calls to
+   * {@link #log()}.
+   */
+  public void log() {
+    m_logCount++;
+    if (m_logWindup == 0 || m_logCount % m_logWindup == 0) {
+      m_stats.update(m_handle);
+      m_output.accept(m_stats);
+    }
+  }
+
+  /**
+   * Sets the log windup value, this value determines how many calls to {@link #log()} are required
+   * before the data is logged. This is available to prevent needless JNI calls when no new data is
+   * available.
+   *
+   * <p>For example, if this value is set to 5, the data will only be logged every 5 calls to {@link
+   * #log()}.
+   *
+   * <p>The {@link PowerDistribution} refreshes it's data every 80ms.
+   *
+   * @param windup The number of calls to {@link #log()} required before the data is logged
+   */
+  public void setLogWindup(int windup) {
+    m_logWindup = windup;
+  }
```

Comment:
This is way premature optimization implemented in a confusing way; updating the log is a handful of JNI functions so it really shouldn't be that much overhead. 

If reducing JNI overhead is a concern the entire thing could be shunted to JNI.

Another interesting option would be to utilize the CAN streaming APIs available for power distribution. This would allow all data to be logged as it arrived, with no wasted log entries.